naptime-testing/src/test/scala/org/coursera/naptime/NestedMacroCourierTests.scala

```diff
@@ -0,0 +1,107 @@
+
+    private[this] def fallbackRelatedFunction(id: String): List[String] = {
+      id match {
+        case "abc" => List("123")
+        case "xyz" => List.empty
+        case "qrs" => List("456", "789")
+        case _ => List.empty
+      }
+    }
+
+    private[this] def makeCourse(id: String): Keyed[String, Course] = {
+      val relatedInstructors = relatedFunction.getOrElse(fallbackRelatedFunction _)(id)
+
+      Keyed(id, Course(s"course-$id", s"$id description", relatedInstructors))
+    }
+
+    def get(id: String) = Nap.get { ctx =>
+      Ok(makeCourse(id))
+    }
+
+    def multiGet(ids: Set[String]) = Nap.multiGet { ctx =>
+      Ok(ids.map(id => makeCourse(id)).toList)
+    }
+
+    def search(query: Option[String]) = Nap.finder { ctx =>
+      val ids = query.map(List(_)).getOrElse(List("abc", "qrs"))
+      Ok(ids.map(id => makeCourse(id)))
+    }
+
+    def byInstructor(instructorId: String) = Nap.finder { ctx =>
+      val ids = instructorId match {
+        case "123" => List("xyz")
+        case "456" => List("xyz", "abc")
+        case "789" => List("qrs")
+        case _ => List.empty
+      }
+      Ok(ids.map(id => makeCourse(id)))
+    }
+  }
+
+  class InstructorsResource extends CourierCollectionResource[String, Instructor] {
+    override def resourceName: String = "instructors"
+
+    def multiGet(ids: Set[String]) = Nap.multiGet { ctx =>
+      ???
+    }
+
+  }
+
+  val courseRouter = Router.build[CoursesResource]
+  val instructorRouter = Router.build[InstructorsResource]
+}
+
+class NestedMacroCourierTests extends AssertionsForJUnit {
+
+  @Test
+  def checkCoursesMergedType(): Unit = {
+    val types = NestedMacroCourierTests.courseRouter.types
+    assert(3 === types.size, s"Got $types")
+    val mergedTypeOpt = types.find(_.key == "org.coursera.naptime.NestedMacroCourierTests.CoursesResource.Model")
+    assert(mergedTypeOpt.isDefined)
+    val mergedType = mergedTypeOpt.get
+    assert(!mergedType.value.hasError)
+    assert(mergedType.value.getType === DataSchema.Type.RECORD)
+    val mergedValueRecord = mergedType.value.asInstanceOf[RecordDataSchema]
+    assert(4 === mergedValueRecord.getFields.size(), mergedValueRecord)
+    val instructorsField = mergedValueRecord.getField("instructors")
+    assert(null != instructorsField, instructorsField)
+    assert(null != instructorsField.getProperties)
+    val typesProperty = instructorsField.getProperties.get(Types.Relations.PROPERTY_NAME)
+    assert(null != typesProperty)
+    assert(typesProperty.isInstanceOf[String])
+    assert(typesProperty === "instructors.v1")
+  }
+
+  @Test
+  def checkMergedCourseModelSchema(): Unit = {
+    val types = NestedMacroCourierTests.courseRouter.types
+    val mergedType = types.find(_.key == "org.coursera.naptime.NestedMacroCourierTests.CoursesResource.Model").get
+    assert(ExpectedMergedCourse.SCHEMA.getFields === mergedType.value.asInstanceOf[RecordDataSchema].getFields)
```

Comment:
I initially wrote the test to compare everything, but the namespaces are off, as well as the type names, so I just check field equality, but I still think that's a step in the right direction. :-)